src/internal/interpreter/http/pipe.ts

```diff
@@ -0,0 +1,504 @@
+import { createUrl, fetchRequest, HttpResponse } from './http';
+import {
+  BINARY_CONTENT_REGEXP,
+  BINARY_CONTENT_TYPES,
+  binaryBody,
+  FetchBody,
+  FetchInstance,
+  FORMDATA_CONTENT,
+  formDataBody,
+  isBinaryBody,
+  isFormDataBody,
+  isStringBody,
+  isUrlSearchParamsBody,
+  JSON_CONTENT,
+  stringBody,
+  URLENCODED_CONTENT,
+  urlSearchParamsBody,
+} from './interfaces';
+import {
+  AuthCache,
+  HttpRequest,
+  ISecurityHandler,
+  RequestParameters,
+} from './security';
+
+/**
+ * Represents input of pipe filter which works with http response
+ */
+export type FetchPipeFilterInput = {
+  parameters: RequestParameters;
+  request: Partial<HttpRequest>;
+  response: HttpResponse | undefined;
+  fetchInstance: FetchInstance & AuthCache;
+  handler: ISecurityHandler | undefined;
+};
+/**
+ * Represents pipe filter which works with http response
+ */
+export type FetchPipeFilter = ({
+  parameters,
+  request,
+  response,
+  fetchInstance,
+  handler,
+}: FetchPipeFilterInput) =>
+  | Pick<FetchPipeFilterInput, 'request' | 'response'>
+  | Promise<Pick<FetchPipeFilterInput, 'request' | 'response'>>;
+
+/**
+ * Represents input of pipe filter which only prepares http request
+ */
+export type PreparePipeFilterInput = {
+  parameters: RequestParameters;
+  request: Partial<HttpRequest>;
+};
+
+/**
+ * Represents pipe filter which only prepares http request
+ * This kind of filter should be able to return same result when ran for the forst time and when ran mutiple times (with same inputs)
+ */
+export type PreparePipeFilter = ({
+  parameters,
+  request,
+}: PreparePipeFilterInput) =>
+  | Pick<PreparePipeFilterInput, 'request'>
+  | Promise<Pick<PreparePipeFilterInput, 'request'>>;
+
+/**
+ * Represents input of pipe which only prepares http request
+ */
+export type PreparePipeInput = {
+  parameters: RequestParameters;
+  fns: PreparePipeFilter[];
+};
+
+/**
+ * Represents output of pipe which only prepares http request
+ */
+export type PreparePipeOutput = HttpRequest;
+
+/**
+ * Represents pipe filter which only prepares http request
+ */
+export type PreparePipe = ({
+  parameters,
+  fns,
+}: PreparePipeInput) => PreparePipeOutput;
+
+/**
+ * Represents pipe filter input which works with http response
+ */
+export type FetchPipeInput = {
+  parameters: RequestParameters;
+  fetchInstance: FetchInstance & AuthCache;
+  handler: ISecurityHandler | undefined;
+  fns: (FetchPipeFilter | PreparePipeFilter)[];
+};
+
+/**
+ * Represents pipe filter output which works with http response
+ */
+export type FetchPipeOutput = Promise<HttpResponse>;
+
+/**
+ * Represents pipe filter which works with http response
+ */
+export type FetchPipe = ({
+  parameters,
+  fns,
+  fetchInstance,
+  handler,
+}: FetchPipeInput) => FetchPipeOutput;
+
+type PipeReturnType<T extends PreparePipeInput | FetchPipeInput> =
+  T extends FetchPipeInput
+    ? HttpResponse
+    : T extends PreparePipeInput
+    ? HttpRequest
+    : never;
```

Comment:
Hmm what I wanted to have was function(pipe) that returns HttpRequest when it gets "prepare" type of filters and HttpResponse when it gets "fetch" type of filters to make it easy to use in both possible use cases.

We could have pipe returning something like: 

{
request: Partial<HttpRequest>,
response: HttpResponse | undefined
}

When we wanted to use request we would have to check if it is complete HttpRequest and if we wanted to use response we would need to check if it is defined. 
It would make typing pipe very simple.

Reply:
Yes, that's what I meant. Request and response both being `{ request: Partial<Request>, response?: Response }`. No need to have multiple "pipe types"